python/lsst/analysis/tools/actions/vector/calcRhoStatistics.py

```diff
@@ -0,0 +1,649 @@
+            xy.process(catA)
+        else:
+            catB = treecorr.Catalog(ra=ra, dec=dec, k=k2, ra_units=raUnits, dec_units=decUnits, logger=_LOG)
+            # Calculate the cross-correlation
+            xy.process(catA, catB)
+
+        _LOG.debug("Correlated %d pairs out of %d possible pairs", xy.npairs, len(ra) * (len(ra) - 1) / 2)
+        return xy
+
+    @classmethod
+    def _corrSpin2(
+        cls,
+        ra: Vector,
+        dec: Vector,
+        g1a: Vector,
+        g2a: Vector,
+        g1b: Vector | None = None,
+        g2b: Vector | None = None,
+        raUnits: str = "degrees",
+        decUnits: str = "degrees",
+        **treecorrKwargs: Any,
+    ) -> GGCorrelation:
+        """Function to compute correlations between shear-like fields.
+
+        This is used to compute Rho statistics, given the appropriate spin-2
+        (shear-like) fields.
+
+        Parameters
+        ----------
+        ra : `numpy.array`
+            The right ascension values of entries in the catalog.
+        dec : `numpy.array`
+            The declination values of entries in the catalog.
+        g1a : `numpy.array`
+            The first component of the primary shear-like field.
+        g2a : `numpy.array`
+            The second component of the primary shear-like field.
+        g1b : `numpy.array`, optional
+            The first component of the secondary shear-like field.
+            Autocorrelation of the primary field is computed if `None`.
+        g2b : `numpy.array`, optional
+            The second component of the secondary shear-like field.
+            Autocorrelation of the primary field is computed if `None`.
+        raUnits : `str`, optional
+            Unit of the right ascension values. Valid options are
+            "degrees", "arcmin", "arcsec", "hours" or "radians".
+        decUnits : `str`, optional
+            Unit of the declination values. Valid options are
+            "degrees", "arcmin", "arcsec", "hours" or "radians".
+        **treecorrKwargs
+            Keyword arguments to be passed to `treecorr.GGCorrelation`.
+
+        Returns
+        -------
+        xy : `treecorr.GGCorrelation`
+            A `treecorr.GGCorrelation` object containing the correlation
+            function.
+        """
+        xy = treecorr.GGCorrelation(**treecorrKwargs)
+        catA = treecorr.Catalog(
+            ra=ra, dec=dec, g1=g1a, g2=g2a, ra_units=raUnits, dec_units=decUnits, logger=_LOG
+        )
+        if g1b is None or g2b is None:
+            # Calculate the auto-correlation
+            xy.process(catA)
+        else:
+            catB = treecorr.Catalog(
+                ra=ra, dec=dec, g1=g1b, g2=g2b, ra_units=raUnits, dec_units=decUnits, logger=_LOG
+            )
+            # Calculate the cross-correlation
+            xy.process(catA, catB)
+
+        _LOG.debug("Correlated %d pairs out of %d possible pairs", xy.npairs, len(ra) * (len(ra) - 1) / 2)
```

Comment:
I would expect this to give an error, although maybe it's just a logging warning.  `xy.npairs` is an array, not an int.  It's the number of pairs for each bin of the output correlation.  It's also a bit weird to report this as being how many were correlated out of some total possible.  The ones that weren't correlated were at different separations, so were they "possible"?

Reply:
Ah, good catch. I meant to put in `sum(xy.npairs)` there.

Reply:
The total number here is to give some indication about whether we have explore sufficient range in the scales or not. As we plan to run this on a regular cadence as we collect data, the angular scales that we set might evolve and this is intended to guide that.

Reply:
I don't think the total of npairs, either as an absolute number or as compared to N*(N-1)/2, would say much about how well you're covering scales though.  Maybe S/N of the first and last bin?  If the S/N of either gets largish, it could mean that it would be useful to extend the range.